### PR TITLE
element-desktop: build native modules

### LINKED ISSUE #87752: Our `element-desktop` is not built with encrypted chat search support
**Describe the bug**

1. Open Riot-Desktop
2. Settings -> Security & Privacy

Under the "Message search" heading, observe:

>Riot is missing some components required for securely caching encrypted messages locally. If you'd like to experiment with this feature, build a custom Riot Desktop with search components added.

it links to here: https://github.com/vector-im/riot-web/blob/develop/docs/native-node-modules.md#adding-seshat-for-search-in-e2e-encrypted-rooms

**Notify maintainers**
cc: @pacien @worldofpeace 

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.6.12, NixOS, 20.09pre-git (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.4`
 - nixpkgs: `/home/cole/code/nixpkgs/cmpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: riot-desktop
```

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

Fixes https://github.com/NixOS/nixpkgs/issues/87752.

Since matrix-seshat has a release with N-API support now, we can completely ignore the hak build system and electron, and link it together afterwards.

Just as the previous solution - this **works on my machines**. If it doesn't work on your machine, please go into as much detail as possible. _I don't plan on debugging this much further if someone just posts "It doesn't work"._

Also, even if some people have problems with this, it might still be better than no support at all.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bb010g** [comment]
Encrypted message search is confirmed working on my NixOS machine via `nix run "github:petabyteboy/nixpkgs/feature/seshat2#element-desktop"`.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 132601` at 55682979 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>element-desktop</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

yarn is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix:294:7:
```
    |
294 |       buildInputs = [ yarn nodejs rsync ] ++ extraBuildInputs;
    |       ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix:334:7:
```
    |
334 |       installPhase = attrs.installPhase or ''
    |       ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

rsync is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix:294:7:
```
    |
294 |       buildInputs = [ yarn nodejs rsync ] ++ extraBuildInputs;
    |       ^
```
</li>
  </ul>
</details>

**Comment by Mic92** [comment]
Same issue as described in https://discourse.nixos.org/t/future-of-npm-packages-in-nixpkgs/14285/13

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 132601` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>element-desktop</li>
  </ul>
</details>

**Review by yayayayaka — COMMENTED** [review]
I'm a bit late to the party but can confirm the application builds on nixos x86_64-linux, indexed my encrypted messages and search results of encrypted messages do appear on my application.

File: pkgs/applications/networking/instant-messengers/element/seshat/default.nix [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  npm_config_nodedir = nodejs-14_x;
+
+  yarnOfflineCache = (callPackage ./yarn.nix {}).offline_cache;
+
+  buildPhase = ''
+    cd ..
+    chmod u+w . ./yarn.lock
+    export HOME=/tmp
+    yarn config --offline set yarn-offline-mirror ${yarnOfflineCache}
+    ${fixup_yarn_lock}/bin/fixup_yarn_lock yarn.lock
```
Comment (andir) [inline pkgs/applications/networking/instant-messengers/element/seshat/default.nix:28]: why isn't `fixup_yarn_lock` in `nativeBuildInputs`?

Comment (yu-re-ka) [reply pkgs/applications/networking/instant-messengers/element/seshat/default.nix:28]: No reason, it should/could be. This is copied from GitLab, which could use the same improvement.

**Pull request merged**